packages/flutter_tools/lib/src/test/runner.dart

```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
 
   // Call package:test's main method in the appropriate directory.
   final Directory saved = fs.currentDirectory;
+  
+  final Function printReloadMessage = () {
+    printStatus(
+      "Press 'r' to hot reload your tests, 'q' to quit",
+      color: TerminalColor.red
+    );
+  };
+
   try {
     if (workDir != null) {
       printTrace('switching to directory $workDir to run tests');
       fs.currentDirectory = workDir;
     }
 
     printTrace('running test package with arguments: $testArgs');
-    await test.main(testArgs);
+    await test.runTests(testArgs);
+
+    if (watchTests) {
+      final Completer<void> completer = Completer<void>();
+      final DirectoryWatcher directoryWatcher = DirectoryWatcher(saved.path);
+
+      directoryWatcher.events.listen((WatchEvent event) {
+        if (event.path.endsWith('.dart')) {
+          invalidatedFiles.add(event.path);
+        }
+      });
+
+      terminal.singleCharMode = true;
+      terminal.onCharInput.asBroadcastStream().listen((String char) async {
+        switch(char) {
+          case 'r':
+            stdout.writeln(terminal.clearScreen());
```

Comment:
I really don't like clearing the terminal here, I lose the context of my previous test run/failure.

Reply:
Removed, I thought it was cleaner to see what the current test state is while running.